File: flow_sdk/builtin/cron_event.py
"""CronEvent entity — scheduled job backed by APScheduler."""

import logging
from datetime import datetime, timezone
from typing import ClassVar, Optional

from starlette.requests import Request

from flow_sdk.api.api_types.api_field import APIField
from flow_sdk.core import action as core_action
from flow_sdk.db.drivers.db_base_record import BuiltinEntityType
from flow_sdk.core.entity.entity_model import Entity
from flow_sdk.request_context.methods import get_current_request_info
from flow_sdk.responses.response import ApiFailResponse, ApiResponse, ApiSuccessResponse

logger = logging.getLogger(__name__)


def _require_desktop():
    from flow_sdk.config import default_service_config
    if not default_service_config.is_desktop:
        from fastapi import HTTPException
        raise HTTPException(status_code=403, detail="CronEvent is only available in desktop mode")


def _parse_trigger(trigger_type: str, expr: str):
    """Parse trigger_type + expr into an APScheduler trigger object."""
    if trigger_type == "interval":
        # expr like "30s", "5m", "2h"
        from apscheduler.triggers.interval import IntervalTrigger
        seconds = _parse_interval_expr(expr)
        return IntervalTrigger(seconds=seconds)
    elif trigger_type == "date":
        from apscheduler.triggers.date import DateTrigger
        run_date = datetime.fromisoformat(expr)
        return DateTrigger(run_date=run_date)
    else:
        # Default: cron
        from apscheduler.triggers.cron import CronTrigger
        return CronTrigger.from_crontab(expr)


def _parse_interval_expr(expr: str) -> int:
    """Convert interval expression like '30s', '5m', '2h' to seconds."""
    expr = expr.strip().lower()
    if expr.endswith("s"):
        return int(expr[:-1])
    elif expr.endswith("m"):
        return int(expr[:-1]) * 60
    elif expr.endswith("h"):
        return int(expr[:-1]) * 3600
    elif expr.endswith("d"):
        return int(expr[:-1]) * 86400
    return int(expr)


async def _fire_cron_job(cron_event_id: str):
    """Callback executed by APScheduler when a cron job fires."""
    try:
        entity = await CronEvent.get_by_id(cron_event_id)
        if entity and entity.enabled:
            entity.counter += 1
            entity.last_run = datetime.now(timezone.utc)
            await entity.update()
            logger.debug(f"CronEvent {entity.name} fired (counter={entity.counter})")
    except Exception as e:
        logger.error(f"CronEvent fire error for {cron_event_id}: {e}")


class CronEvent(Entity):
    """Scheduled job entity backed by APScheduler."""

    type: str = APIField(default=BuiltinEntityType.CRON_EVENT.value)
    name: str = APIField()
    description: Optional[str] = APIField(None)
    expr: str = APIField(default="* * * * *")
    trigger_type: str = APIField(default="cron")
    enabled: bool = APIField(default=True)
    counter: int = APIField(default=0)
    last_run: Optional[datetime] = APIField(None)
    next_run: Optional[datetime] = APIField(None)

    _api_visible: ClassVar[bool] = True
    _unique: ClassVar[list[str]] = []

    # ── Read (list / get-by-id) ──────────────────────────────────────────────

    @core_action.get(action_name="read")
    async def list_action(cls, request: Request) -> ApiResponse:
        """
        GET /api/v1/graph/cron_event         → sync APScheduler → DB, return all
        GET /api/v1/graph/cron_event/{id}    → get by ID (also syncs next_run)
        """
        _require_desktop()
        request_info = get_current_request_info()

        # Single entity fetch
        if request_info and request_info.target_entity_typeid and request_info.target_entity_typeid.id:
            entity = await CronEvent.get_by_id(request_info.target_entity_typeid.id)
            if entity is None:
                return ApiFailResponse(message="CronEvent not found")
            await _sync_entity_from_scheduler(entity)
            return ApiSuccessResponse(data=entity)

        # List — sync all scheduler jobs → DB first
        await _sync_all_from_scheduler()
        entities = await CronEvent.get_all()
        return ApiSuccessResponse(data=entities)

    # ── Create ───────────────────────────────────────────────────────────────

    @core_action.post(action_name="create")
    async def create_action(cls, request: Request) -> ApiResponse:
        """POST /api/v1/graph/cron_event"""
        _require_desktop()
        request_info = get_current_request_info()
        body = await request_info.get_post_data() if request_info else {}
        if not body:
            return ApiFailResponse(message="Request body required")

        name = body.get("name")
        if not name:
            return ApiFailResponse(message="name is required")

        expr = body.get("expr", "* * * * *")
        trigger_type = body.get("trigger_type", "cron")
        description = body.get("description")
        enabled = body.get("enabled", True)

        entity = CronEvent(
            name=name,
            description=description,
            expr=expr,
            trigger_type=trigger_type,
            enabled=enabled,
        )
        await entity.save()

        # Register with APScheduler
        try:
            scheduler = _get_scheduler()
            if scheduler:
                trigger = _parse_trigger(trigger_type, expr)
                job = scheduler.add_job(
                    _fire_cron_job,
                    trigger=trigger,
                    id=entity.id,
                    name=name,
                    args=[entity.id],
                    replace_existing=True,
                )
                if not enabled:
                    job.pause()
                # Capture next_run
                if job.next_run_time:
                    entity.next_run = job.next_run_time
                    await entity.update()
        except Exception as e:
            logger.warning(f"Failed to schedule cron job: {e}")

        return ApiSuccessResponse(data=entity)

    # ── Update ───────────────────────────────────────────────────────────────

    @core_action.all(action_name="update")
    async def update_action(self, request: Request) -> ApiResponse:
        """PUT/PATCH /api/v1/graph/cron_event/{id}"""
        _require_desktop()
        request_info = get_current_request_info()
        body = await request_info.get_post_data() if request_info else {}
        if not body:
            return ApiFailResponse(message="Request body required")

        # Apply updates to entity fields
        for field in ("name", "description", "expr", "trigger_type", "enabled"):
            if field in body:
                setattr(self, field, body[field])

        await self.update()

        # Reschedule in APScheduler
        try:
            scheduler = _get_scheduler()
            if scheduler:
                trigger = _parse_trigger(self.trigger_type, self.expr)
                job = scheduler.reschedule_job(self.id, trigger=trigger)
                if job:
                    if self.enabled:
                        job.resume()
                    else:
                        job.pause()
                    if job.next_run_time:
                        self.next_run = job.next_run_time
                        await self.update()
        except Exception as e:
            logger.warning(f"Failed to reschedule cron job {self.id}: {e}")

        return ApiSuccessResponse(data=self)

    # ── Delete ───────────────────────────────────────────────────────────────

    @core_action.delete(action_name="delete")
    async def delete_action(self, request: Request) -> ApiResponse:
        """DELETE /api/v1/graph/cron_event/{id}"""
        _require_desktop()
        # Remove from APScheduler
        try:
            scheduler = _get_scheduler()
            if scheduler:
                scheduler.remove_job(self.id)
        except Exception as e:
            logger.debug(f"APScheduler remove_job error (may not exist): {e}")

        await self.delete()
        return ApiSuccessResponse(data={"deleted": True})

    # ── Test ─────────────────────────────────────────────────────────────────

    @core_action.post(action_name="test")
    async def test_action(self, request: Request) -> ApiResponse:
        """POST /api/v1/graph/cron_event/{id}/test — fire the job immediately."""
        _require_desktop()
        self.counter += 1
        self.last_run = datetime.now(timezone.utc)
        await self.update()
        return ApiSuccessResponse(data={"status": "fired", "counter": self.counter})


# ── Helpers ──────────────────────────────────────────────────────────────────

def _get_scheduler():
    """Get running scheduler, or None if unavailable."""
    try:
        from flow_sdk.server.scheduler import get_scheduler
        return get_scheduler()
    except Exception:
        return None


async def _sync_entity_from_scheduler(entity: CronEvent):
    """Update a single entity's next_run from APScheduler state."""
    try:
        scheduler = _get_scheduler()
        if scheduler and entity.id:
            job = scheduler.get_job(entity.id)
            if job:
                new_next = job.next_run_time
                if new_next != entity.next_run:
                    entity.next_run = new_next
                    await entity.update()
    except Exception as e:
        logger.debug(f"Scheduler sync error for {entity.id}: {e}")


async def _sync_all_from_scheduler():
    """Sync all APScheduler jobs → DB entities."""
    try:
        scheduler = _get_scheduler()
        if scheduler is None:
            return

        jobs = scheduler.get_jobs()
        job_ids = {job.id for job in jobs}

        # Update next_run for existing jobs
        for job in jobs:
            entity = await CronEvent.get_by_id(job.id)
            if entity:
                entity.next_run = job.next_run_time
                await entity.update()

        # Remove DB entities whose scheduler job no longer exists
        all_entities = await CronEvent.get_all()
        for entity in all_entities:
            if entity.id and entity.id not in job_ids:
                # Orphaned — remove from DB
                await entity.delete()
    except Exception as e:
        logger.debug(f"Full scheduler sync error: {e}")

File: ts_sdk/src/entities/cron-event.ts
import type { EntityMerge } from '../IEntity';
import { APIEntity, dataManager, registerEntity } from '../APIEntity';
import { QueryRequest } from '../FlowSync/query';
import { IEntity } from '../IEntity';
import { ActionInfo } from '../models/ActionInfo';
import { HttpMethod } from '../models/ApiUrl';

export interface ICronEvent extends IEntity {
  name: string;
  description?: string;
  expr: string;
  trigger_type?: string;
  enabled?: boolean;
  counter?: number;
  last_run?: Date;
  next_run?: Date;
}

// `implements ICronEvent` only checks the class; it contributes no members, so every
// field declared solely on ICronEvent read as "does not exist". deepAssign populates
// them from the wire — this merge makes them part of the class type.
// eslint-disable-next-line @typescript-eslint/no-empty-object-type
export interface CronEvent extends EntityMerge<ICronEvent> {}

/**
 * Scheduled job entity backed by APScheduler
 */
@registerEntity
export class CronEvent extends APIEntity<CronEvent> implements ICronEvent {
  static type: string = 'cron_event';

  name: string = '';
  description?: string;
  expr: string = '* * * * *';
  trigger_type: string = 'cron';
  enabled: boolean = true;
  counter: number = 0;
  last_run?: Date;
  next_run?: Date;

  constructor(entity: Partial<ICronEvent> = {}) {
    super(entity);
    this.name = entity.name || '';
    this.description = entity.description;
    this.expr = entity.expr || '* * * * *';
    this.trigger_type = entity.trigger_type || 'cron';
    this.enabled = entity.enabled !== undefined ? entity.enabled : true;
    this.counter = entity.counter ?? 0;
    this.last_run = entity.last_run;
    this.next_run = entity.next_run;
  }

  /**
   * List all cron events (triggers APScheduler sync on backend)
   */
  static async list(): Promise<CronEvent[]> {
    const request = new QueryRequest({
      type: CronEvent.type,
      query: null,
      scope: [],
    });
    return await CronEvent.query(request, true);
  }

  /**
   * Create a new cron event
   */
  static async create(data: Partial<ICronEvent>): Promise<CronEvent> {
    const action = new ActionInfo('create', CronEvent.type, null, 'POST' as HttpMethod);
    action.bodyParameters = data as Record<string, unknown>;
    const response = await dataManager.callAction<Record<string, unknown>, ICronEvent>(action);
    return new CronEvent(response as ICronEvent);
  }

  /**
   * Update this cron event
   */
  async updateFields(data: Partial<ICronEvent>): Promise<CronEvent> {
    if (!this.id) {
      throw new Error('Cannot update unsaved CronEvent');
    }
    const action = new ActionInfo('update', CronEvent.type, this.id, 'PUT' as HttpMethod);
    action.bodyParameters = data as Record<string, unknown>;
    const response = await dataManager.callAction<Record<string, unknown>, ICronEvent>(action);
    return new CronEvent(response as ICronEvent);
  }

  /**
   * Delete this cron event
   */
  async remove(): Promise<void> {
    if (!this.id) {
      throw new Error('Cannot delete unsaved CronEvent');
    }
    const action = new ActionInfo('delete', CronEvent.type, this.id, 'DELETE' as HttpMethod);
    await dataManager.callAction(action);
  }

  /**
   * Fire this cron job immediately (test mode)
   */
  async test(): Promise<{ status: string; counter: number }> {
    if (!this.id) {
      throw new Error('Cannot test unsaved CronEvent');
    }
    const action = new ActionInfo('test', CronEvent.type, this.id, 'POST' as HttpMethod);
    const response = await dataManager.callAction<undefined, { status: string; counter: number }>(action);
    return response as { status: string; counter: number };
  }
}
